cassandra/cluster.py

```diff
@@ -3783,6 +3783,141 @@ def _refresh_schema(self, connection, preloaded_results=None, schema_agreement_w
         self._cluster.metadata.refresh(connection, self._timeout, fetch_size=self._schema_meta_page_size, **kwargs)
 
         return True
+    
+    # Three functions below (refresh_schema_async, _refresh_schema_async, _wait_for_schema_agreement_async) are async
+    # versions of the functions without _async in name - instead of blocking and returning result, their first argument
+    # is a callback that will receive either a result or an exception.
+    # Purpose of those functions is to avoid filling whole thread pool and deadlocking.
+    
+    # INTENDED ONLY FOR INTERNAL USE
+    def refresh_schema_async(self, callback, force=False, **kwargs):
+        def new_callback(e):
+            if isinstance(e, ReferenceError):
+                # our weak reference to the Cluster is no good
+                callback(False)
+                return
+            elif isinstance(e, Exception):
+                log.debug("[control connection] Error refreshing schema", exc_info=True)
+                self._signal_error()
+                callback(False)
+                return
+            else:
+                callback(e)
+        if self._connection:
+            self._refresh_schema_async(new_callback, self._connection, force=force, **kwargs)
+        else:
+            callback(False)
+
+    # INTENDED ONLY FOR INTERNAL USE
+    def _refresh_schema_async(self, callback, connection, preloaded_results=None, schema_agreement_wait=None, force=False, **kwargs):
+        if self._cluster.is_shutdown:
+            callback(False)
+            return
+        
+        def new_callback(e):
+            if not self._schema_meta_enabled and not force:
+                log.debug("[control connection] Skipping schema refresh because schema metadata is disabled")
+                callback(False)
+                return
+
+            if not e:
+                log.debug("Skipping schema refresh due to lack of schema agreement")
+                callback(False)
+                return
+            self._cluster.metadata.refresh(connection, self._timeout, fetch_size=self._schema_meta_page_size, **kwargs)
+        
+        self._wait_for_schema_agreement_async(new_callback,
+                                                connection=self._connection,
+                                                preloaded_results=preloaded_results,
+                                                wait_time=schema_agreement_wait)
+
+    # INTENDED ONLY FOR INTERNAL USE
+    def _wait_for_schema_agreement_async(self, callback, connection=None, preloaded_results=None, wait_time=None):
```

Comment:
Doesn't the function name starting with `_` already suggest that it's `INTENDED ONLY FOR INTERNAL USE`?